zeppelin-web/src/components/notebookListDataFactory/notebookList.datafactory.js

```diff
@@ -17,10 +17,47 @@ angular.module('zeppelinWebApp').factory('notebookListDataFactory', function() {
   var notes = {};
 
   notes.list = [];
+  notes.flatList = [];
 
   notes.setNotes = function(notesList) {
-    notes.list = angular.copy(notesList);
+    notes.flatList = angular.copy(notesList); // a flat list to boost searching
```

Comment:
I'm not convinced that representing a hierarchy with lists is such a good idea. Why not use objects instead? Then we can rely on lodash's [set](https://lodash.com/docs#set) and [get](https://lodash.com/docs#get), like this:

```javascript
// set notes
    var notes = notesList.reduce(function(acc, note) {
      var noteName = note.name || note.id;
      var dirs = noteName.match(/([^\\\][^\/]|\\\/)+/g);

      return _.set(
        acc,
        dirs.join("."),
        {id: note.id}
      )
    }, {})

// get a note at "/C/CB/CBA"
  _.get(notes, "C.CB.CBA")

```

What do you think?